src/handlers/storage-handler.ts

```diff
@@ -1,76 +1,132 @@
 import { isFunction, strEqualsIgnoreCase } from '../utils/types'
-import { StorageStrategy } from '../model/storage-strategy'
-import { EventBus, ExternalStorageHandler, IStorageHandler, StorageStrategyMode } from '../types'
+import { StorageStrategies, StorageStrategy } from '../model/storage-strategy'
+import { EventBus, ExternalMinimalStorageHandler, ExternalStorageHandler } from '../types'
+import { WrappingContext } from '../utils/wrapping'
 
-const _noOp = () => undefined
+interface WrappedExternalMinimalStorageHandler {
+  getCookie: (key: string) => string | null | undefined,
+  getDataFromLocalStorage: (key: string) => string | null | undefined,
+  localStorageIsEnabled: () => boolean | undefined,
+}
 
-export function StorageHandler (storageStrategy: StorageStrategyMode, externalStorageHandler: ExternalStorageHandler, eventBus: EventBus): IStorageHandler {
-  const errors = []
+interface WrappedExternalStorageHandler {
+  setCookie: (key: string, value: string, expires?: Date, sameSite?: string, domain?: string) => void,
+  setDataInLocalStorage: (key: string, value: string) => void,
+  removeDataFromLocalStorage: (key: string) => void,
+  findSimilarCookies: (substring: string) => string[] | undefined
+}
 
-  function _externalOrError (functionName: string): CallableFunction {
-    const hasExternal = externalStorageHandler && externalStorageHandler[functionName] && isFunction(externalStorageHandler[functionName])
-    if (strEqualsIgnoreCase(storageStrategy, StorageStrategy.disabled)) {
-      return _noOp
-    } else if (hasExternal) {
-      return externalStorageHandler[functionName].bind(externalStorageHandler)
-    } else {
-      errors.push(functionName)
-      return _noOp
+const noop = () => undefined
+
+function wrapRead<T extends object, K extends keyof T & string> (wrapper: WrappingContext<T>, storageStrategy: StorageStrategy, functionName: K) {
```

Comment:
I would rename it to something else as it is called also in the wrapWriter
should we call it`wrap` ?

Reply:
Write is generally stronger than read, so IMO it feels fine to implement write permission checks in terms of read. Other opinions?

Reply:
You're right +1